src/Microsoft.Health.Fhir.SqlServer/Features/Storage/SqlServerFhirDataStore.cs

```diff
@@ -91,443 +84,251 @@ internal class SqlServerFhirDataStore : IFhirDataStore, IProvideCapability
-                        //// TODO: remove _mergeResourcesRetriesFlag when set bundle processing is in place.
-                        || (_mergeResourcesRetriesFlag.IsEnabled() && e.IsRetriable()) // this should allow to deal with intermittent database errors.
-                        || ((isExecutionTimeout = _mergeResourcesRetriesFlag.IsEnabled() && e.IsExecutionTimeout()) && retries++ < 3)) // timeouts happen once in a while on highly loaded databases.
+                    var isConflict = sqlEx != null && sqlEx.Number == SqlErrorCodes.Conflict;
+                    if (isConflict && retries++ < 30) // retries on conflict should never be more than 1, so it is OK to hardcode.
+                    //// TODO: we cannot retry today as this call might be in outer transaction. Uncomment 2 lines below when set bundle processing is in place.
+                    //// || (isRetriable = e.IsRetriable()) // this should allow to deal with intermittent database errors.
+                    //// || (isExecutionTimeout = e.IsExecutionTimeout()) && retries++ < 3) // timeouts happen once in a while on highly loaded databases.
                     {
                         _logger.LogWarning(e, $"Error from SQL database on {nameof(MergeAsync)} retries={{Retries}}", retries);
-                        if (isConflict || isExecutionTimeout)
+                        if (isRetriable || isExecutionTimeout) // others are logged in SQL by merge stored procedure
                         {
-                            await TryLogEvent(nameof(MergeAsync), "Warn", $"Error={e.Message}, retries={retries}", mergeStart, cancellationToken);
+                            await TryLogEvent(nameof(MergeAsync), "Warn", $"retries={retries}, error={e}, ", mergeStart, cancellationToken);
                         }
 
                         await Task.Delay(5000, cancellationToken);
                         continue;
                     }
 
                     _logger.LogError(e, $"Error from SQL database on {nameof(MergeAsync)} retries={{Retries}}", retries);
-                    await TryLogEvent(nameof(MergeAsync), "Error", $"Error={e.Message}, retries={retries}", mergeStart, cancellationToken);
-                    throw;
+                    if (sqlEx != null)
+                    {
+                        await TryLogEvent(nameof(MergeAsync), "Error", $"retries={retries}, error={sqlEx}", mergeStart, cancellationToken);
+                    }
+
+                    throw trueEx;
                 }
             }
         }
 
-        private async Task TryLogEvent(string process, string status, string text, DateTime? startDate, CancellationToken cancellationToken)
+        // split in a separate method to allow special logic in $import
+        internal async Task<Dictionary<ResourceKey, UpsertOutcome>> MergeMainAsync(IReadOnlyList<ResourceWrapperOperation> resources, CancellationToken cancellationToken)
```

Comment:
Small suggestion: I've seen the dotnet framework team naming this type of method as "internals". 
Then, in this case, the suggestion would be to rename from "MergeMainAsync" to "MergeInternalAsync".

Reply:
I will change.